python/phase-2.md

```diff
@@ -0,0 +1,830 @@
+# Python SDK - Phase 2
+
+The Python SDK will be implemented in three loosely defined phases:
+
+* Phase 1 - Initial project, activity worker, and client
+* Phase 2 - Workflow worker, local activities, async activity completion, etc
+* Phase 3 - Deterministic workflow sandbox, replayer, test suite (some parts may be combined with phase 2)
+
+This proposal is for phase 2. It is intentionally loosely defined and all aspects are subject to discussion/change.
+
+## Overview
+
+Following up from phase 1 and known implementation changes since then, this phase will implement the actual workflow
+worker/code and other advanced features.
+
+A simple example of a workflow definition and worker run:
+
+```python
+from datetime import timedelta
+from temporalio import activity, workflow
+from temporalio.client import Client
+from temporalio.worker import Worker
+
+@activity.defn
+async def say_hello(name: str) -> str:
+    return f"Hello, {name}!"
+
+@workflow.defn
+class MyWorkflow:
+    @workflow.run
+    async def run(arg1: str) -> str:
+        return await workflow.execute_local_activity(
+            say_hello, arg1, start_to_close_timeout=timedelta(seconds=5)
+        )
+
+
+async def main(stop_event: asyncio.Event):
+    client = await Client.connect("http://localhost:7233")
+    async with Worker(
+        client,
+        task_queue="somequeue",
+        activities=[say_hello],
+        workflows=[MyWorkflow],
+    ):
+        await stop_event.wait()
+```
+
+## Workflow Definitions
+
+A workflow is a class.
+
+Here is an example of a class-based workflow that calls a local activity to get
+a greeting with a signal to update the prefix and a query to get the latest
+greeting:
+
+```python
+import asyncio
+from datetime import timedelta
+from temporalio import activity, workflow
+from temporalio.client import Client
+from temporalio.worker import Worker
+
+@activity.defn
+async def create_greeting_activity(salutation: str, name: str) -> str:
+    return f"{salutation}, {name}!"
+
+@workflow.defn
+class GreetingWorkflow:
+    def __init__() -> None:
+        self._salutation = "Hello"
+        self._salutation_update = asyncio.Event()
+        self._complete = asyncio.Event()
+
+    @workflow.run
+    async def run(self, name: str) -> str:
+        while True:
+            # Store greeting (this is typed)
+            self._current_greeting = await workflow.execute_local_activity(
+                create_greeting_activity,
+                self._salutation,
+                name,
+                start_to_close_timeout=timedelta(seconds=5),
+            )
+            workflow.logger.debug("Greeting set to %s", self._current_greeting)
```

Comment:
How / where did you set the log destination?

Reply:
This is a Python thing. See https://docs.python.org/3/howto/logging.html, https://docs.python.org/3/howto/logging-cookbook.html, https://docs.python.org/3/library/logging.html, etc. Think of it like slf4j+logback or log4j in Java. Temporal SDKs provide a workflow logger that is smart enough to not log during replay.